Add instructions for running a migration from Whitehall

brucebolt opened the pull request:
These include instructions for running the migration locally without needing to replicate data.

Trello card: https://trello.com/c/W6Ay4rYb

kevindew commented on docs/import-from-whitehall.md:
````diff
@@ -0,0 +1,110 @@
+
+```
+Edition.find(1034451).document.id
+```
````
Might be good to reference that you have to be using a Whitehall Rails console and it needs to be in the environment. We could also give a command that takes the whole URL rather than giving people instructions of what to pull out the URL. For example:

```
irb(main):005:0> URI.parse("https://whitehall-admin.integration.publishing.service.gov.uk/government/admin/publications/1038279").path.split("/")[4].then { |id| Edition.find(id).document_id }
=> 412144
```

brucebolt replied:
Updated the text above to say it must be a Whitehall Rails console.  Also changed the code to parse a URL.

kevindew commented on docs/import-from-whitehall.md:
```diff
@@ -0,0 +1,110 @@
+
+## Running a migration locally using govuk-docker
+
+Using [govuk-docker](https://github.com/alphagov/govuk-docker), you can run an import locally without needing to replicate data.  In order to do this, you will need to override the [Plek](https://github.com/alphagov/plek) URL for Whitehall and Publishing API to point to staging.  Additionally, you will require the bearer tokens for these applications from [govuk-secrets](https://github.com/alphag ...
```
Is there a particular reason we suggest staging rather than integration?

brucebolt replied:
I was previously having trouble accessing Publishing API in integration.

However, it appears I should've been using https://publishing-api.blue.integration.govuk.digital rather than https://publishing-api.integration.publishing.service.gov.uk.  Have updated the document to now use integration.

kevindew commented on docs/import-from-whitehall.md:
```diff
@@ -0,0 +1,110 @@
+
+The values you need to obtain are:
+- `govuk::apps::content_publisher::whitehall_bearer_token` - the Whitehall bearer token
+- `govuk::apps::content_publisher::publishing_api_bearer_token` - the Publishing API bearer token
```
We shouldn't really be pulling out these tokens as these are the production ones and we should be really cautious that these could accidentally be leaked if used as part of dev.

What I'd suggest we should do is that rather than pulling out tokens, developers should be creating their own tokens from signon. This can be done by going to: https://signon.integration.publishing.service.gov.uk/api_users and going to create new API user.

brucebolt replied:
That's a good point.  I've change this to give instructions for creating an API user and getting that user's tokens.

kevindew commented on docs/import-from-whitehall.md:
````diff
@@ -0,0 +1,110 @@
+
+```
+govuk-docker run -e PLEK_SERVICE_PUBLISHING_API_URI=https://publishing-api.staging.publishing.service.gov.uk -e PUBLISHING_API_BEARER_TOKEN=<> -e PLEK_SERVICE_WHITEHALL_ADMIN_URI=https://whitehall-admin.staging.publishing.service.gov.uk -e WHITEHALL_BEARER_TOKEN=<> content-publisher-worker bundle exec sidekiq -C config/sidekiq.yml
+```
````
This looks really helpful 👍

kevindew commented on docs/import-from-whitehall.md:
````diff
@@ -0,0 +1,110 @@
+govuk-docker run -e PLEK_SERVICE_PUBLISHING_API_URI=https://publishing-api.staging.publishing.service.gov.uk -e PUBLISHING_API_BEARER_TOKEN=<> -e PLEK_SERVICE_WHITEHALL_ADMIN_URI=https://whitehall-admin.staging.publishing.service.gov.uk -e WHITEHALL_BEARER_TOKEN=<> content-publisher-worker bundle exec sidekiq -C config/sidekiq.yml
+```
+
+Replace `<>` with the bearer tokens you have previously obtained.
````
I wonder if it'd be a bit clearer if we used example vars like $whitehall_bearer_token and $publishing_api_bearer_token.

Could also tell users how to export them first with 

```
export whitehall_bearer_token=<your token>
export publishing_api_bearer_token=<your token>
```

and then the longer commands could be pasted verbatim

brucebolt replied:
Updated throughout.  Also setting the URLs in the same way to reduce the length of the commands.

kevindew commented on docs/import-from-whitehall.md:
````diff
@@ -0,0 +1,110 @@
+Edition.find(1034451).document.id
+```
+
+### Import documents of a specific type (and optional subtypes) from a single publishing organisation
````
I feel like we should suggest that people pass in the subtypes by default rather than saying it's optional.

This seems to have caught Emma out (see https://trello.com/c/nibnOHkx/1538-dont-export-government-responses-from-whitehall) where I've suggested we might want to do more with defaults.

brucebolt replied:
I've removed the words optional.  We can update this if any changes are made to the optional nature of the subtypes.

kevindew commented on docs/import-from-whitehall.md:
````diff
@@ -0,0 +1,110 @@
+
+```
+govuk-docker-run rake import:whitehall_migration["nuclear-decommissioning-authority","news_article"] env PLEK_SERVICE_PUBLISHING_API_URI=https://publishing-api.staging.publishing.service.gov.uk PUBLISHING_API_BEARER_TOKEN=<> PLEK_SERVICE_WHITEHALL_ADMIN_URI=https://whitehall-admin.staging.publishing.service.gov.uk WHITEHALL_BEARER_TOKEN=<>
+```
````
There's a few inconsistencies with this command with the previous one that seem like something that could catch someone out. We need to be in a specific directory in this example and do environment variables differently.

Might be good to switch it to be consistent, for example:

```
govuk-docker run -e PLEK_SERVICE_PUBLISHING_API_URI=https://publishing-api.staging.publishing.service.gov.uk -e PUBLISHING_API_BEARER_TOKEN=<> -e PLEK_SERVICE_WHITEHALL_ADMIN_URI=https://whitehall-admin.staging.publishing.service.gov.uk -e WHITEHALL_BEARER_TOKEN=<> bundle exec rake import:whitehall_migration["nuclear-decommissioning-authority","news_article"]
```

brucebolt replied:
Changed this to be consistent with the other commands.

kevindew commented on docs/import-from-whitehall.md:
````diff
@@ -0,0 +1,110 @@
+In order to access the URL printed at the end of the rake task, the Content Publisher application will need to be running locally:
+
+```
+govuk-docker-up
````
This one is also directory specific could switch it to `govuk-docker up content-publisher-app`

brucebolt replied:
Changed this one too.

brucebolt commented on docs/import-from-whitehall.md:
```diff
@@ -0,0 +1,110 @@
+
+There are two ways to import documents from Whitehall:
+- a single document
+- documents of a specific type or subtypes from a single publishing organisation
```
Yes, I've added internal links

brucebolt commented:
Thank you for the review @kevindew.  I've addressed all your comments, and worked out how to get the import working against the integration environment.

kevindew commented on docs/import-from-whitehall.md:
````diff
@@ -0,0 +1,136 @@
+```
+export publishing_api_url=https://publishing-api.staging.publishing.service.gov.uk
+export whitehall_url=https://whitehall-admin.staging.publishing.service.gov.uk
+```
````
Does the `blue.staging.govuk.digital` work for this? it seems rather weird to be using the carrenza URL here and not sure if it's intended to be supported or not. Life might be simpler if we can use the govuk.digital one in case it's a mistake that this different one is available.

brucebolt replied:
Unfortunately not.  This can be seen by visiting https://publishing-api.staging.publishing.service.gov.uk/healthcheck which gives a 200, whereas http://publishing-api.blue.staging.govuk.digital/healthcheck times out.

kevindew replied:
If you use the https equivalent of the latter it works.

brucebolt replied:
Ah, I didn't see that.  Have updated.

kevindew commented on docs/import-from-whitehall.md:
```diff
@@ -0,0 +1,136 @@
+
+### Accessing a Rails console
+
+You may wish to open a Rails console that allows you to perform tasks on Whitehall documents (e.g. unlock or lock a document in order to test the import task multiple times).
```
You might want to link to https://github.com/alphagov/whitehall/blob/master/docs/migration_to_content_publisher/locked-documents.md for the "unlock or lock documents" since that is potentially unfamiliar terminology.

brucebolt replied:
Added a link to this doc
